src/ledger/P23HotArchiveBug.cpp

```diff
@@ -451,5 +451,211 @@ Protocol23CorruptionDataVerifier::verifyEntryFixesOnP24Upgrade(
+    {
+        LedgerEntry corruptedEntry =
+            decodeLedgerEntry(P23_CORRUPTED_HOT_ARCHIVE_ENTRIES[i]);
+
+        LedgerEntry correctEntry =
+            decodeLedgerEntry(P23_CORRUPTED_HOT_ARCHIVE_ENTRY_CORRECT_STATE[i]);
+
+        mKeyToEntries.emplace(LedgerEntryKey(corruptedEntry),
+                              std::make_pair(correctEntry, corruptedEntry));
+    }
+}
+
+std::optional<std::pair<uint64_t /*lo balance*/, SCAddress /*owner*/>>
+getAmountLo(LedgerEntry const& le)
+{
+    if (le.data.type() != CONTRACT_DATA)
+    {
+        return std::nullopt;
+    }
+    auto const& cd = le.data.contractData();
+
+    if (cd.durability != ContractDataDurability::PERSISTENT ||
+        cd.key.type() != SCV_VEC || !cd.key.vec() || cd.key.vec()->size() != 2)
+    {
+        return std::nullopt;
+    }
+
+    SCVal balanceSymbol(SCV_SYMBOL);
+    balanceSymbol.sym() = "Balance";
+
+    // The balanceSymbol should be the first entry in the SCVec
+    if (!(cd.key.vec()->at(0) == balanceSymbol))
+    {
+        return std::nullopt;
+    }
+
+    auto const& balanceOwner = cd.key.vec()->at(1);
+    if (balanceOwner.type() != SCV_ADDRESS)
+    {
+        return std::nullopt;
+    }
+
+    if (cd.val.type() != SCV_MAP || !cd.val.map() || cd.val.map()->size() != 3)
+    {
+        return std::nullopt;
+    }
+
+    SCVal amountSymbol(SCV_SYMBOL);
+    amountSymbol.sym() = "amount";
+
+    auto const& amountEntry = cd.val.map()->at(0);
+    if (!(amountEntry.key == amountSymbol) ||
+        amountEntry.val.type() != SCV_I128)
+    {
+        return std::nullopt;
+    }
+
+    // This looks like an SAC balance entry.
+    auto lo = amountEntry.val.i128().lo;
+    auto hi = amountEntry.val.i128().hi;
+    // Make the assumption that amount fits in int64_t for now. We'll find out
+    // if this is false during validation. Remember that this will only run for
+    // p23 after we're on p24.
+    releaseAssert(hi == 0);
+
+    // We checked earlier that the key is a vector of size 2.
+    return std::make_pair(lo, balanceOwner.address());
+}
+
+std::optional<Protocol23CorruptionEventReconciler::SACReconciliationInfo>
+Protocol23CorruptionEventReconciler::getSACReconciliationEventAndTrackDiff(
+    LedgerKey const& restoredKey, LedgerEntry const& restoredEntry,
+    uint32_t ledgerSeq, uint32_t protocolVersion)
+{
+    if (!protocolVersionEquals(protocolVersion, ProtocolVersion::V_23))
+    {
+        return std::nullopt;
+    }
+
+    auto entryIter = mKeyToEntries.find(restoredKey);
+    if (entryIter == mKeyToEntries.end())
+    {
+        return std::nullopt;
+    }
+
+    auto restoreLo = getAmountLo(restoredEntry);
+    if (!restoreLo)
+    {
+        return std::nullopt;
+    }
+
+    auto const& [correctEntry, corruptedEntry] = entryIter->second;
+    releaseAssert(restoredEntry == corruptedEntry);
+
+    auto correctLo = getAmountLo(correctEntry);
+    if (!correctLo)
+    {
+        CLOG_WARNING(Ledger,
```

Comment:
This should never be the case, right?

Reply:
I changed to assert that the balance exists. I wasn't sure if none of the SAC instances were restored, which is why the check was a little more general. I'll change this to be more strict and try it out in parallel catchup.

Reply:
Makes sense, I'm pretty sure there were no corrupted instances, but there could be one with the lastModifiedLedgerSeq corrupted (I haven't looked at those entries)